### PR TITLE
Project limits

### LINKED ISSUE #6169: Limits (quotas) for projects
Projects as they are today are a great way to segment a LXD host, having its own containers, images and profiles (depending on configuration). Combined with RBAC, it's then possible to have some users/groups only have access to specific projects.

But right now, access to a project means you still get to create as many containers as you want, use as much CPU and memory as you want, potentially running the host system out of resources.

We shouldn't try to fix all of this in one shot, but need to put some infrastructure in place for such limits to be implemented. The easiest would be to use the existing project configuration mechanism, adding more configuration keys like:

 - limits.containers (total number of containers that can be held by the project)
 - limits.cpu (maximum number of CPUs that can be assigned to a container OR range of CPUs that may be used)
 - limits.memory (total amount of memory usable by the project)
 - limits.processes (total number of processes usable by the project)

I suggest we do this initial set as a way to prove the concept, we can then add more as we see demand.

### PR DESCRIPTION
**Pull request by freeekanayaka**
Fixes #6169.

Signed-off-by: Free Ekanayaka <free.ekanayaka@canonical.com>

### CODE REVIEW

**Comment by stgraber** [comment]
s/dectect/detect/

**Comment by freeekanayaka** [comment]
> s/dectect/detect/

Fixed.

**Comment by freeekanayaka** [comment]
Ok, this ready for review. I hope it covers the intended semantics of limits.

File: lxd/containers_post.go [hunk 1]
```diff
@@ -13,7 +13,7 @@ import (
 	"os"
 	"strings"
 
-	"github.com/dustinkirkland/golang-petname"
+	petname "github.com/dustinkirkland/golang-petname"
```
Comment (stgraber) [inline lxd/containers_post.go]: emacs having fun

File: lxd/containers_post.go [hunk 2]
```diff
@@ -827,6 +801,57 @@ func containersPost(d *Daemon, r *http.Request) response.Response {
...
+	//
+	// Note this check is performed after automatically generated config
+	// values (such as the ones from an InstanceType) have been set.
+	err = d.cluster.Transaction(func(tx *db.ClusterTx) error {
+		if req.Type == "" {
+			switch req.Source.Type {
+			case "copy":
+				if req.Source.Source == "" {
+					return fmt.Errorf("Must specify a source instance")
+				}
```
Comment (stgraber) [inline lxd/containers_post.go]: line break

File: lxd/containers_post.go [hunk 3]
```diff
@@ -827,6 +801,57 @@ func containersPost(d *Daemon, r *http.Request) response.Response {
...
+				source, err := instance.LoadInstanceDatabaseObject(tx, project, req.Source.Source)
+				if err != nil {
+					return errors.Wrap(err, "Load source instance from database")
+				}
```
Comment (stgraber) [inline lxd/containers_post.go]: line break

File: lxd/containers_post.go [hunk 4]
```diff
@@ -827,6 +801,57 @@ func containersPost(d *Daemon, r *http.Request) response.Response {
...
+				req.Type = api.InstanceType(source.Type.String())
+			case "migration":
+				req.Type = api.InstanceTypeContainer
+			}
+		}
+		err := projecthelpers.CheckLimitsUponInstanceCreation(tx, project, req)
+		if err != nil {
+			return err
+		}
```
Comment (stgraber) [inline lxd/containers_post.go]: line break

File: lxd/containers_post.go [hunk 5]
```diff
@@ -827,6 +801,57 @@ func containersPost(d *Daemon, r *http.Request) response.Response {
...
+				}
+				source, err := instance.LoadInstanceDatabaseObject(tx, project, req.Source.Source)
+				if err != nil {
+					return errors.Wrap(err, "Load source instance from database")
+				}
+				req.Type = api.InstanceType(source.Type.String())
+			case "migration":
+				req.Type = api.InstanceTypeContainer
+			}
+		}
```
Comment (stgraber) [inline lxd/containers_post.go]: line break

File: lxd/containers_post.go [hunk 6]
```diff
@@ -827,6 +801,57 @@ func containersPost(d *Daemon, r *http.Request) response.Response {
...
+		err := projecthelpers.CheckLimitsUponInstanceCreation(tx, project, req)
+		if err != nil {
+			return err
+		}
+		if req.Name == "" {
+			names, err := tx.InstanceNames(project)
+			if err != nil {
+				return err
+			}
```
Comment (stgraber) [inline lxd/containers_post.go]: line break

File: lxd/api_project.go [hunk 7]
```diff
@@ -327,6 +328,13 @@ func projectChange(d *Daemon, project *api.Project, req api.ProjectPut) response
 			configChanged = append(configChanged, key)
 		}
 	}
```
Comment (stgraber) [inline lxd/api_project.go]: line break

File: lxd/project/limits.go [hunk 8]
```diff
@@ -397,6 +399,16 @@ var aggregateLimitConfigValueParsers = map[string]func(string) (int64, error){
 		}
 		return int64(limit), nil
 	},
+	"limits.cpu": func(value string) (int64, error) {
+		if strings.Contains(value, ",") || strings.Contains(value, "-") {
+			return -1, fmt.Errorf("CPUs can't be pinned if project limits are used")
+		}
```
Comment (stgraber) [inline lxd/project/limits.go]: line break

File: lxd/project/limits.go [hunk 9]
```diff
@@ -397,6 +399,16 @@ var aggregateLimitConfigValueParsers = map[string]func(string) (int64, error){
...
+		limit, err := strconv.Atoi(value)
+		if err != nil {
+			return -1, err
+		}
```
Comment (stgraber) [inline lxd/project/limits.go]: line break

**Comment by stgraber** [comment]
Looks good minus the few listed nits.

**Comment by freeekanayaka** [comment]
Nits should be addressed.

**Pull request merged**